protoarray/src/main/java/tech.pegasys.artemis.protoarray/ProtoArrayForkChoice.java

```diff
@@ -0,0 +1,292 @@
+    try {
+      List<UnsignedLong> oldBalances = balances;
+      List<UnsignedLong> newBalances = justifiedStateBalances;
+
+      List<Long> deltas = computeDeltas(protoArray.getIndices(), votes, oldBalances, newBalances);
+
+      protoArray.applyScoreChanges(deltas, justifiedEpoch, finalizedEpoch);
+      balances = newBalances;
+
+      return protoArray.findHead(justifiedRoot);
+    } finally {
+      protoArrayLock.writeLock().unlock();
+      votesLock.writeLock().unlock();
+      balancesLock.writeLock().unlock();
+    }
+  }
+
+  public void maybePrune(Bytes32 finalizedRoot) {
+    protoArrayLock.writeLock().lock();
+    try {
+      protoArray.maybePrune(finalizedRoot);
+    } finally {
+      protoArrayLock.writeLock().unlock();
+    }
+  }
+
+  public void setPruneThreshold(int pruneThreshold) {
+    protoArrayLock.writeLock().lock();
+    try {
+      protoArray.setPruneThreshold(pruneThreshold);
+    } finally {
+      protoArrayLock.writeLock().unlock();
+    }
+  }
+
+  public int size() {
+    protoArrayLock.readLock().lock();
+    try {
+      return protoArray.getNodes().size();
+    } finally {
+      protoArrayLock.readLock().unlock();
+    }
+  }
+
+  public boolean containsBlock(Bytes32 blockRoot) {
+    protoArrayLock.readLock().lock();
+    try {
+      return protoArray.getIndices().containsKey(blockRoot);
+    } finally {
+      protoArrayLock.readLock().unlock();
+    }
+  }
+
+  public Optional<BlockSlotAndStateRoot> blockSlotAndStateRoot(Bytes32 blockRoot) {
+    protoArrayLock.readLock().lock();
+    try {
+      int blockIndex = protoArray.getIndices().get(blockRoot);
+
+      if (protoArray.getNodes().size() > blockIndex) {
+        ProtoNode node = protoArray.getNodes().get(blockIndex);
+        return Optional.of(new BlockSlotAndStateRoot(node.getSlot(), node.getStateRoot()));
+      } else {
+        return Optional.empty();
+      }
+    } finally {
+      protoArrayLock.readLock().unlock();
+    }
+  }
+
+  public Optional<Checkpoint> latestMessage(int validatorIndex) {
+    votesLock.readLock().lock();
+    try {
+      if (validatorIndex < votes.size()) {
+        VoteTracker vote = votes.get(validatorIndex);
+        if (vote.equals(VoteTracker.DEFAULT)) {
+          return Optional.empty();
+        } else {
+          return Optional.of(new Checkpoint(vote.getNextEpoch(), vote.getNextRoot()));
+        }
+      } else {
+        return Optional.empty();
+      }
+    } finally {
+      votesLock.readLock().unlock();
+    }
+  }
+
+  /**
+   * Returns a list of `deltas`, where there is one delta for each of the indices in
+   * `0..indices.size()`.
+   *
+   * The deltas are formed by a change between `oldBalances` and `newBalances`,
+   * and/or a change of vote in `votes`.
+   *
+   * ## Errors
+   *
+   * - If a value in `indices` is greater to or equal to `indices.size()`.
+   * - If some `Bytes32` in `votes` is not a key in `indices` (except for `Bytes32.ZERO`, this is
+   * always valid).
```

Comment:
Spotless is going to do bad things to this...  It will work out better as
```suggestion
      * <ul>
   *   <li>If a value in `indices` is greater to or equal to `indices.size()`.
   *   <li>If some `Bytes32` in `votes` is not a key in `indices` (except for `Bytes32.ZERO`, this
   *       is always valid).
   * </ul>
```

because JavaDoc is HTML.

Reply:
Interesting. Good to know. Done.